### PR TITLE
feat: WDS button group widget integration

### LINKED ISSUE #27748: Integrate WDS ButtonGroup component with appsmith
- [Example of button integration](https://github.com/appsmithorg/appsmith/pull/25106)
- [ButtonGroup in Storybrook](https://release--64c89da059b7d62ad3a60544.chromatic.com/?path=/docs/design-system-widgets-button-group--docs)
- [ButtonGroup code](https://github.com/appsmithorg/appsmith/tree/release/app/client/packages/design-system/widgets/src/components/ButtonGroup)

In order to get acquainted with what properties should be in property pane, please read [this doc](https://www.notion.so/appsmith/Widgets-Properties-Audit-3832f6a7be5f47cc9bfa3bd1ddae31fa).

### PR DESCRIPTION
**Pull request by dhruvikn**
## Description
Adds WDSButtonGroup Widget.

#### PR fixes following issue(s)
Fixes #27748

#### Type of change
- Chore (housekeeping or task changes that don't impact user perception)

- [ ] Manual
- [ ] JUnit
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

**Comment by dhruvikn** [comment]
Please note that I have pulled changes from [`feat/wds-icon-button-widget`](https://github.com/appsmithorg/appsmith/pull/27928) in this branch. Hence, please ignore the files in `app/client/src/widgets/wds/WDSIconButtonWidget/*` directory while reviewing. @KelvinOm @jsartisan

**Review by KelvinOm — COMMENTED** [review]
I don't really understand why we need to disable buttonGroup itself, besides, this option probably doesn't work right now. Let's delete it?
Oct-19-2023 11-18-11

File: app/client/src/widgets/wds/WDSIconButtonWidget/widget/index.tsx
[the remark this answers is not in the document]
Comment (dhruvikn) [reply app/client/src/widgets/wds/WDSIconButtonWidget/widget/index.tsx:1]: Yes, I [dropped a comment here earlier](https://github.com/appsmithorg/appsmith/pull/28205#issuecomment-1769333785).

File: app/client/src/widgets/wds/WDSButtonGroupWidget/component/index.tsx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+          button.iconName &&
+          (() => {
+            return <BIcon icon={button.iconName} />;
+          });
+
+        const hasOnClickAction = () => {
+          return Boolean(button.onClick && !button.isDisabled);
+        };
+
+        const handleActionComplete = () => {
```
[the remark this answers is not in the document]
Comment (dhruvikn) [reply app/client/src/widgets/wds/WDSButtonGroupWidget/component/index.tsx:41]: We show a loading indicator when an action is ongoing. This callback removes the button ID from our loadingButtonIds  state.

**Review by KelvinOm — APPROVED** [review]

**Review by KelvinOm — DISMISSED** [review]

**Comment by shadabbuchh** [comment]
Bugs in this PR:

1. On DnDing the widget, it does not occupy the whole space within the size handlers.
Screenshot 2023-10-20 at 5 10 31 PM

2. Issues in the `Style` tab:

https://github.com/appsmithorg/appsmith/assets/39921438/6d4ea6f9-1e80-44e6-ab1d-070f214cb61a

3. On DnDing an icon button when we already have Button group in the canvas changes the color of the icons.
4. No tooltip/title appears when name of the button is long. Same goes for when we have multiple buttons.
Screenshot 2023-10-20 at 5 28 21 PM

5. Icon colors change when clicking on the JS toggle button (Similar to what happens in Icon button).
6. When a button is not visible, the size handlers do not re-arrange, instead we get to see a lot of empty space.

https://github.com/appsmithorg/appsmith/assets/39921438/f68d7d58-9c94-4319-9bc2-d1c6b25240e2

7. When a button is disabled we don't get to see the appropriate cursor.

**Comment by dhruvikn** [comment]
@shadabbuchh - 

> 1. On DnDing the widget, it does not occupy the whole space within the size handlers.

As discussed, we are not fixing anything related to layouts in the widgets at the moment.

> 2. Issues in the `Style` tab:

1. Have fixed the variant labels.
2. To track the flickering of icons problem, I have created a ticket here - https://github.com/appsmithorg/appsmith/issues/28460
3. Not fixing the issue where you were able to dnd an icon button over the button group in the vertical orientation because we are not fixing anything related to layouts in the widgets at the moment. It should be resolved with Anvil.

> 3. On DnDing an icon button when we already have Button group in the canvas changes the color of the icons.

This is again the due to the same issue where icon color is getting overridden when opening the property pane. Tracking it here - https://github.com/appsmithorg/appsmith/issues/28353

> 4. No tooltip/title appears when name of the button is long. Same goes for when we have multiple buttons.

Have created a separate issue for this - https://github.com/appsmithorg/appsmith/issues/28459.

> 5. Icon colors change when clicking on the JS toggle button (Similar to what happens in Icon button).

This is due to the same flicker issue occurring in point number 2. Should be fixed in this  - https://github.com/appsmithorg/appsmith/issues/28460

> 6. When a button is not visible, the size handlers do not re-arrange, instead we get to see a lot of empty space.

As discussed, we are not fixing anything related to layouts in the widgets at the moment.

> 7. When a button is disabled we don't get to see the appropriate cursor.

Have added cursor not allowed when hovering over a disabled button.

**Comment by shadabbuchh** [comment]
Issues that are fixed have been retested. Looks good.

**Pull request merged**